packages/datadog-instrumentations/src/couchbase.js

```diff
@@ -0,0 +1,155 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource,
+  bindEmit
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+addHook({ name: 'couchbase', file: 'lib/bucket.js', versions: ['^2.6.5'] }, Bucket => {
+  const startChn1qlReq = channel('apm:couchbase:_n1qlReq:start')
+  const asyncEndChn1qlReq = channel('apm:couchbase:_n1qlReq:async-end')
+  const endChn1qlReq = channel('apm:couchbase:_n1qlReq:end')
+  const errorChn1qlReq = channel('apm:couchbase:_n1qlReq:error')
+
+  Bucket.prototype._maybeInvoke = wrapMaybeInvoke(Bucket.prototype._maybeInvoke)
+  Bucket.prototype.query = wrapQuery(Bucket.prototype.query)
+
+  shimmer.wrap(Bucket.prototype, '_n1qlReq', _n1qlReq => function (host, q, adhoc, emitter) {
+    if (
+      !startChn1qlReq.hasSubscribers
+    ) {
+      return _n1qlReq.apply(this, arguments)
+    }
+    if (!emitter || !emitter.once) return _n1qlReq.apply(this, arguments)
+    const n1qlQuery = q && q.statement
+
+    startChn1qlReq.publish([n1qlQuery, this])
+
+    emitter.once('rows', AsyncResource.bind(() => {
+      asyncEndChn1qlReq.publish(undefined)
+    }))
+
+    emitter.once('error', AsyncResource.bind((error) => {
+      errorChn1qlReq.publish(error)
+      asyncEndChn1qlReq.publish(undefined)
+    }))
+
+    try {
+      return _n1qlReq.apply(this, arguments)
+    } catch (err) {
+      err.stack // trigger getting the stack at the original throwing point
+      errorChn1qlReq.publish(err)
+
+      throw err
+    } finally {
+      endChn1qlReq.publish(undefined)
+    }
+  })
+
+  Bucket.prototype.upsert = wrap('apm:couchbase:upsert', Bucket.prototype.upsert)
+  Bucket.prototype.insert = wrap('apm:couchbase:insert', Bucket.prototype.insert)
+  Bucket.prototype.replace = wrap('apm:couchbase:replace', Bucket.prototype.replace)
+  Bucket.prototype.append = wrap('apm:couchbase:append', Bucket.prototype.append)
+  Bucket.prototype.prepend = wrap('apm:couchbase:prepend', Bucket.prototype.prepend)
+
+  return Bucket
+})
+
+addHook({ name: 'couchbase', file: 'lib/cluster.js', versions: ['^2.6.5'] }, Cluster => {
+  Cluster.prototype._maybeInvoke = wrapMaybeInvoke(Cluster.prototype._maybeInvoke)
+  Cluster.prototype.query = wrapQuery(Cluster.prototype.query)
+
+  shimmer.wrap(Cluster.prototype, 'openBucket', openBucket => function (name, callback) {
+    const ar = new AsyncResource('bound-anonymous-fn')
+    const res = openBucket.apply(this, arguments)
+    bindEmit(res, ar)
```

Comment:
Which events specifically need to be bound? Also, why do they need to be bound and could the surrounding code be bound instead?

Reply:
For that eventemitter, the events "connect" and "error" need to be bound